### PR TITLE
Test JSON NeXus with ScippNexus

### LINKED ISSUE #454: Refactor JSON NeXus tests to use ScippNexus
#359 discusses removal of `load_nexus`. For loading HDF5 NeXus files this poses no problem as unit test in ScippNexus cover this. However, ScippNeutron contains a large number of tests for `load_nexus` that operate also on JSON NeXus, using an `h5py` adapter. We need to preserve those tests to ensure that loading JSON NeXus keeps working with ScippNexus.

It is not entirely clear if ScippNeutron is the correct place for the `h5py` adapter and these tests.

### PR DESCRIPTION
**Pull request by jl-wynen**
Fixes #454

I only translated a small number of test cases because the majority of old tests are now covered by ScippNexus. The new tests only check that `JSONGroup` works as a replacement for `h5.Group`. Please check if I missed some case!

The type hints in ScippNexus need to be updated for this. Or we make `JSONGroup` inherit from `h5.Group`.

### CODE REVIEW

**Comment by SimonHeybrock** [comment]
> The type hints in ScippNexus need to be updated for this. Or we make `JSONGroup` inherit from `h5.Group`.

Why? What needs updating? I thought ScippNexus is using some `Protocol` classes for this?

**Comment by jl-wynen** [comment]
> > The type hints in ScippNexus need to be updated for this. Or we make `JSONGroup` inherit from `h5.Group`.
> 
> Why? What needs updating? I thought ScippNexus is using some `Protocol` classes for this?

You are right. `JSONGroup` does not implement the protocol. And doing so does not make much sense because we don't need methods for writing data. But I can create those methods and simply raise `NotImplementedError`.

File: tests/io/load_nexus_json_test.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
...
+    builder.add_stream(Stream(stream_path))
+    builder.add_dataset_at_path("/entry/source/depends_on", stream_path, {})
+
+    with pytest.warns(UserWarning):
+        loaded_data, _ = load_nexus_json_str(builder.json_string)
+
+    # A 0 distance translation is used in place of the streamed transformation
+    default = [0, 0, 0]
+    assert np.allclose(loaded_data["source_position"].values, default)
+    assert loaded_data["source_position"].unit == sc.Unit("m")
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply tests/io/load_nexus_json_test.py:62]: This is an old test. I only moved the file which is why it shows up in the diff.

File: tests/io/json_nexus_examples/event_data.json [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+          "string_size": 2,
+          "type": "string",
+          "name": "units",
+          "values": "ns"
+        },
+        {
+          "string_size": 20,
+          "type": "string",
+          "name": "offset",
+          "values": "1970-01-01T00:00:00Z"
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply tests/io/json_nexus_examples/event_data.json:74]: Good idea!

File: tests/io/load_nexus_json_test.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+    binned = sc.bins(data=expected_events, dim='event', begin=begin, end=end)
+    expected = sc.DataArray(
+        binned.rename_dims({'event': 'event_time_zero'}),
+        coords={'event_time_zero': event_time_zero},
+    )
+
+    sc.testing.assert_identical(loaded.coords['event_time_zero'], event_time_zero)
+    # Using `loaded == expected` could fail if the order of events in the
+    # bins is different.
+    # Since the order is arbitrary, check that the bins have equal weights instead.
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply tests/io/load_nexus_json_test.py:113]: Is this guaranteed by the API?

File: tests/io/json_nexus_examples/log.json [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+          "string_size": 1,
+          "type": "string",
+          "name": "units",
+          "values": "s"
+        },
+        {
+          "string_size": 20,
+          "type": "string",
+          "name": "start",
+          "values": "1970-01-01T00:00:00Z"
```
[the remark this answers is not in the document]
Comment (jl-wynen) [reply tests/io/json_nexus_examples/log.json:53]: It looks like it is not handled.
Is `"start"` the correct name for the attribute? In the even data, it is called `"offset"`.

[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply tests/io/json_nexus_examples/log.json:53]: `start` should be correct: https://manual.nexusformat.org/classes/base_classes/NXlog.html

File: tests/io/load_nexus_json_test.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply tests/io/load_nexus_json_test.py:113]: Yes.

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**